tsdb/shard_test.go

```diff
@@ -1696,6 +1697,45 @@ func TestMeasurementFieldSet_InvalidFormat(t *testing.T) {
 	}
 }
 
+func TestMeasurementFieldSet_ConcurrentSave(t *testing.T) {
+	var iterations int64
+	dir, cleanup := MustTempDir()
+	defer cleanup()
+
+	if testing.Short() {
+		iterations = 50
+	} else {
+		iterations = 200
+	}
+
+	path := filepath.Join(dir, "fields.idx")
+	mf, err := tsdb.NewMeasurementFieldSet(path)
+	if err != nil {
+		t.Fatalf("NewMeasurementFieldSet error: %v", err)
+	}
+	var wg sync.WaitGroup
+
+	wg.Add(4)
+	go testFieldMaker(t, &wg, mf, "cpu", "cvalue", iterations)
+	go testFieldMaker(t, &wg, mf, "cpu", "cvalue", iterations)
+	go testFieldMaker(t, &wg, mf, "cpu", "cvalue", iterations)
+	go testFieldMaker(t, &wg, mf, "cpu", "cvalue", iterations)
+	wg.Wait()
+}
+
+func testFieldMaker(t *testing.T, wg *sync.WaitGroup, mf *tsdb.MeasurementFieldSet, measurement string, fieldName string, count int64) {
+	defer wg.Done()
+	fields := mf.CreateFieldsIfNotExists([]byte(measurement))
+	for ; count > 0; count -= 1 {
+		if err := fields.CreateFieldIfNotExists([]byte("value"+strconv.FormatInt(count, 10)), influxql.Float); err != nil {
+			t.Fatalf("create field error: %v", err)
+		}
+		if err := mf.Save(); err != nil {
+			t.Fatalf("save error: %v", err)
+		}
+	}
+}
```

Comment:
Does this change improve the benchmark for this? I assume this one:

`func BenchmarkWritePoints_NewSeries_1000000_Measurement_1_TagKey_1_TagValue(b *testing.B) {`

Reply:
Should also check the rest of the benchmarks in the same file to see if there are regressions I think.

Reply:
Will check!

Reply:
The benchmarks don't really cover this case.  The only one close is BenchmarkWritePoints_ExistingSeries_100K_1_1. We would need a benchmark that just created millions of measurements and fields, and wrote them without batching. The new code and the existing code in master-1.x do not complete the benchmarks, and both timeout at the 11 minute mark. There does not appear to be any regression in the benchmarks, but since they already have errors, it's hard to be sure.

Reply:
I'd like some proof that this will actually improve performance - is there anything we can do to show that?

Reply:
How difficult would it be to fix or add benchmarks to measure the change in performance for this change? If possible, it would be best to get some measurements since this is in the storage layer.